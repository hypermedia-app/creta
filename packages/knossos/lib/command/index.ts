export { init } from './init'
export { serve } from './serve'
