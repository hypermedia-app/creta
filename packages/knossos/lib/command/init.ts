import { promisify } from 'util'
import path from 'path'
import * as fs from 'fs-extra'

export async function init(): Promise<void> {
  const sourceDir = path.resolve(__dirname, '../resources')
  const destDir = path.resolve(process.cwd(), 'resources')

  return promisify(fs.copy)(sourceDir, destDir)
}
