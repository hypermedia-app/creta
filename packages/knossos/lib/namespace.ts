import namespace from '@rdfjs/namespace'

type KnossosTerms = 'createWithPUT' | 'memberTemplate' | 'SystemAccount'

export const knossos = namespace<KnossosTerms>('https://hypermedia.app/knossos#')
