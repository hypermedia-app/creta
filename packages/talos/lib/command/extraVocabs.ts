import { ExtraVocab } from '../insertVocabs'

export function parseExtraVocabs(value: string, vocabs: ExtraVocab[] = []): ExtraVocab[] {
  const [packageName, ...prefixes] = value.split(',')

  return [
    ...vocabs,
    {
      package: packageName,
      prefixes: prefixes.map(p => p.trim()).filter(p => !!p),
    },
  ]
}
