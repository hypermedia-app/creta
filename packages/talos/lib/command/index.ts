export { put } from './put'
export { putVocabs } from './put-vocabs'

export interface Command {
  endpoint: string
  updateEndpoint?: string
  token?: string
  user?: string
  password?: string
}
