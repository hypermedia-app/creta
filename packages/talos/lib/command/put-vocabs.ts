import $rdf from 'rdf-ext'
import { ExtraVocab, insertVocabs } from '../insertVocabs'
import { deleteApi } from '../deleteApi'
import { log } from '../log'
import type { Command } from '.'

export interface PutVocabs extends Command {
  apiDoc?: string
  extraVocabs?: Array<ExtraVocab>
}

export async function putVocabs({ token, endpoint, user, password, extraVocabs, apiDoc }: PutVocabs) {
  await insertVocabs({
    endpointUrl: endpoint,
    updateUrl: endpoint,
    user,
    password,
  }, {
    extraVocabs,
  }).then(() => log('Inserted vocabularies'))

  if (apiDoc) {
    await deleteApi({
      apiUri: $rdf.namedNode(apiDoc),
      token,
    })
  }
}
