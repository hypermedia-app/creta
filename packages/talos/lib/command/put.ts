import path from 'path'
import debug, { Debugger } from 'debug'
import fetch from 'node-fetch'
import { insertVocabs } from '../insertVocabs'
import { bootstrap } from '../bootstrap'

interface Put {
  api: string
  endpoint: string
  user?: string
  password?: string
  dir: string
  vocabs?: boolean
  resources?: boolean
  token?: string
}

async function insertResources({ dir, token, api, endpoint, user, password }: Put, log: Debugger) {
  const cwd = path.resolve(process.cwd(), dir)

  await bootstrap({
    log,
    api,
    cwd,
    endpointUrl: endpoint,
    updateUrl: endpoint,
    user,
    password,
  })

  if (token) {
    const res = await fetch(`${api}/api`, {
      method: 'DELETE',
      headers: {
        Authorization: `System ${token}`,
      },
    })

    if (res.ok) {
      log('Reset hydra:ApiDocumentation')
    } else {
      log('Failed to reset hydra:ApiDocumentation: %s', await res.text())
    }
  } else {
    log('No System token provided. API restart may be necessary for changes to be applied')
  }
}

export async function put(arg: Put): Promise<void> {
  const log = debug('talos')
  log.enabled = true

  let inserted = false

  if (arg.vocabs) {
    inserted = true
    await insertVocabs({
      endpointUrl: arg.endpoint,
      updateUrl: arg.endpoint,
      user: arg.user,
      password: arg.password,
    }).then(() => log('Inserted vocabularies'))
  }

  if (arg.resources) {
    inserted = true
    await insertResources(arg, log)
  }

  if (!inserted) {
    log('Nothing selected for bootstrapping. Please check --help option')
  }
}
