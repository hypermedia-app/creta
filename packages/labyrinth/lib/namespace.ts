import namespace from '@rdfjs/namespace'

type AuthTerms = 'required' | 'scopes' | 'permissions' | 'access'

export const hydraBox = namespace('http://hydra-box.org/schema/')
export const query = namespace('https://hypermedia.app/query#')
export const auth = namespace<AuthTerms>('https://hypermedia.app/auth#')
export const code = namespace('https://code.described.at/')
export const acl = namespace('http://www.w3.org/ns/auth/acl#')
