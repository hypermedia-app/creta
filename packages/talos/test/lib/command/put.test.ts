import * as path from 'path'
import { put, Put } from 'talos/lib/command/put'
import { ASK, DELETE } from '@tpluscode/sparql-builder'
import ParsingClient from 'sparql-http-client/ParsingClient'
import { expect } from 'chai'
import { doap, hydra, schema } from '@tpluscode/rdf-ns-builders'
import namespace from '@rdfjs/namespace'
import * as NodeFetch from 'node-fetch'
import sinon from 'sinon'

const api = 'http://example.com/base'
const ns = namespace(api + '/')
const dir = path.resolve(__dirname, '../../resources')

describe('@hydrofoil/talos/lib/command/put', () => {
  const params: Put = {
    api,
    endpoint: 'http://db.labyrinth.lndo.site/repositories/tests',
    user: 'minos',
    password: 'password',
  }

  const client = new ParsingClient({
    endpointUrl: 'http://db.labyrinth.lndo.site/repositories/tests',
    updateUrl: 'http://db.labyrinth.lndo.site/repositories/tests',
    user: 'minos',
    password: 'password',
  })

  before(async () => {
    await DELETE`?s ?p ?o`.WHERE`?s ?p ?o`.execute(client.query)
  })

  describe('--resources', () => {
    before(async () => {
      await put([dir], params)
    })

    describe('turtle', () => {
      it('inserts into graph constructed from path', () => {
        const userCreated = ASK`${ns('project/creta/user/tpluscode')} a ${schema.Person}`
          .FROM(ns('project/creta/user/tpluscode'))
          .execute(client.query)

        expect(userCreated).to.eventually.be.true
      })

      it('escapes paths to produce valid URIs', () => {
        const userCreated = ASK`${ns('project/creta/user/Kov%C3%A1cs%20J%C3%A1nos')} a ${schema.Person}`
          .FROM(ns('project/creta/user/Kov%C3%A1cs%20J%C3%A1nos'))
          .execute(client.query)

        expect(userCreated).to.eventually.be.true
      })

      it('adds apiDocumentation link', () => {
        const userCreated = ASK`${ns('project/creta/user/tpluscode')} ${hydra.apiDocumentation} ?api`
          .FROM(ns('project/creta/user/tpluscode'))
          .execute(client.query)

        expect(userCreated).to.eventually.be.true
      })
    })

    describe('n-quads', () => {
      it('inserts into graph constructed from path', () => {
        const userCreated = ASK`${ns('project/creta/project/creta')} a ${doap.Project}`
          .FROM(ns('project/creta/project/creta'))
          .execute(client.query)

        expect(userCreated).to.eventually.be.true
      })
    })
  })

  describe('--resources --token', () => {
    let fetch: sinon.SinonStub

    before(async () => {
      fetch = sinon.stub(NodeFetch, 'default').resolves({
        ok: true,
      } as any)

      await put([dir], {
        ...params,
        token: 'foo-bar',
        apiPath: '/da-api',
      })
    })

    afterEach(() => {
      sinon.restore()
    })

    it('requests DELETE of ApiDocumentation', () => {
      // expect
      expect(fetch).to.have.been.calledWith('http://example.com/base/da-api', sinon.match({
        method: 'DELETE',
        headers: {
          Authorization: sinon.match(/^System .+/),
        },
      }))
    })
  })
})
