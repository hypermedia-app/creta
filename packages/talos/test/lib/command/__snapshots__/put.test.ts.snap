exports["@hydrofoil/talos/lib/command/put --resources api http://example.com turtle replaces entire graph by default"] = "<http://example.com/project> <http://schema.org/parentItem> <http://example.com> .\n<http://example.com/project> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Thing> .\n<http://example.com/project> <http://www.w3.org/ns/hydra/core#apiDocumentation> <http://example.com/api> .\n";

exports["@hydrofoil/talos/lib/command/put --resources api http://example.com turtle merge existing graph when annotated"] = "<http://example.com/project/creta/user.group/admins> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2006/vcard/ns#Group> .\n<http://example.com/project/creta/user.group/admins> <http://www.w3.org/2006/vcard/ns#hasMember> <http://example.com/project/creta/user/tpluscode> .\n<http://example.com/project/creta/user.group/admins> <http://www.w3.org/2006/vcard/ns#n> \"Administrators\" .\n<http://example.com/project/creta/user.group/admins> <http://www.w3.org/ns/hydra/core#apiDocumentation> <http://example.com/api> .\n";

exports["@hydrofoil/talos/lib/command/put --resources api http://example.com/base turtle replaces entire graph by default"] = "<http://example.com/base/project> <http://schema.org/parentItem> <http://example.com/base> .\n<http://example.com/base/project> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Thing> .\n<http://example.com/base/project> <http://www.w3.org/ns/hydra/core#apiDocumentation> <http://example.com/base/api> .\n";

exports["@hydrofoil/talos/lib/command/put --resources api http://example.com/base turtle merge existing graph when annotated"] = "<http://example.com/base/project/creta/user.group/admins> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2006/vcard/ns#Group> .\n<http://example.com/base/project/creta/user.group/admins> <http://www.w3.org/2006/vcard/ns#hasMember> <http://example.com/base/project/creta/user/tpluscode> .\n<http://example.com/base/project/creta/user.group/admins> <http://www.w3.org/2006/vcard/ns#n> \"Administrators\" .\n<http://example.com/base/project/creta/user.group/admins> <http://www.w3.org/ns/hydra/core#apiDocumentation> <http://example.com/base/api> .\n";

